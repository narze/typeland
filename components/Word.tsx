/** @jsx jsx */
import { css, jsx } from '@emotion/core'
import tw from '@tailwindcssinjs/macro'
import React, { useRef, useEffect, useState } from 'react'

const s = {
  word: tw`
    relative
    mr-1
  `,
  correct: tw`
    text-green-600
  `,
  wrong: tw`
    text-red-400
  `,
  pending: tw`
    text-gray-300
  `,
  caret: [
    tw`
      absolute
      h-6
      bg-blue-500
    `,
    css`
      transition: all 0.1s ease-out;
      width: 2px;
    `,
  ],
}

export interface WordProps {
  template: string
  userInput: string
  showCaret: boolean
}

export const Word: React.FC<WordProps> = React.memo(
  ({ template, userInput, showCaret }) => {
    const lastInputEl = useRef(null)
    const [caretPos, setCaretPos] = useState({
      left: -1,
      top: 4,
    })

    const caretPosCss = () => {
      return css`
        left: ${caretPos.left + 1}px;
        top: ${caretPos.top}px;
      `
    }

    useEffect(() => {
      if (!lastInputEl.current) {
        setCaretPos((prev) => {
          return {
            ...prev,
            left: 0,
          }
        })
        return
      }

      setCaretPos((prev) => {
        return {
          ...prev,
          left:
            lastInputEl.current.offsetLeft + lastInputEl.current.offsetWidth,
        }
      })
    }, [userInput.length])

    return (
      <span css={s.word} data-testid="word">
        {Array.from(
          Array(Math.max(template.length, userInput.length)).keys()
        ).map((i) => {
          const ref = i == userInput.length - 1 ? lastInputEl : null
          const templateChar = template[i]
          const userInputChar = userInput[i]

          let charElement
          if (templateChar && userInputChar) {
            charElement =
              templateChar == userInputChar ? (
                <span
                  css={s.correct}
                  className="correct"
                  data-testid="correct"
                  ref={ref}
                >
                  {templateChar}
                </span>
              ) : (
                <span
                  css={s.wrong}
                  className="wrong"
                  data-testid="wrong"
                  ref={ref}
                >
                  {templateChar}
                </span>
              )
          } else if (userInputChar) {
            charElement = (
              <span
                css={s.wrong}
                className="wrong"
                data-testid="wrong"
                ref={ref}
              >
                {userInputChar}
              </span>
            )
          } else {
            charElement = (
              <span
                css={s.pending}
                className="pending"
                data-testid="pending"
                ref={ref}
              >
                {templateChar}
              </span>
            )
          }

          return <React.Fragment key={i}>{charElement}</React.Fragment>
        })}

        {showCaret && (
          <span css={[s.caret, caretPosCss()]} data-testid="caret"></span>
        )}
      </span>
    )
  }
)
