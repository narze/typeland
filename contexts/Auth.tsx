import React, { useEffect, useReducer } from 'react'
import { auth as firebaseAuth } from '../config/firebase'
import firebase from 'firebase'

interface AuthState {
  user: firebase.User | null
}

interface AuthContextProps {
  state: AuthState
  dispatch: React.Dispatch<any>
}

interface AuthProviderProps {
  children: React.ReactNode
  value?: AuthContextProps
}

const defaultAuth: AuthState = {
  user: null,
}

type Action = { type: 'setUser'; user: firebase.User } | { type: 'unsetUser' }

export const reducer = (state: AuthState, action: Action): AuthState => {
  switch (action.type) {
    case 'setUser':
      return {
        ...state,
        user: action.user,
      }
    case 'unsetUser':
      return {
        ...state,
        user: null,
      }
    default:
      return state
  }
}

export const AuthContext = React.createContext({} as AuthContextProps)

export const AuthProvider: React.FC<AuthProviderProps> = ({
  children,
  value,
}) => {
  const [state, dispatch] = useReducer(reducer, value.state || defaultAuth)

  useEffect(() => {
    const unsubscribe = firebaseAuth.onAuthStateChanged((authUser) => {
      if (authUser) {
        //The user is logged in
        dispatch({
          type: 'setUser',
          user: authUser,
        })
      } else {
        //The user is logged out
        dispatch({
          type: 'unsetUser',
        })
      }
    })
    return () => {
      // Any clean up operation goes in here
      unsubscribe()
    }
  }, [])

  return (
    <AuthContext.Provider value={{ state, dispatch }}>
      {children}
    </AuthContext.Provider>
  )
}
