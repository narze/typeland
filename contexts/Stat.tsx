import React from 'react'

const stats = {
  correct: 0,
  wrong: 0,
  total: 0,
}

export const StatContext = React.createContext(stats)
