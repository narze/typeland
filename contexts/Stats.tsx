import React, { useContext, useReducer } from 'react'

interface State {
  correct: number
  wrong: number
  total: number
}

type Dispatch = React.Dispatch<Action>

interface StatsProviderProps {
  children: React.ReactNode
  initialState?: State
}

export type Action =
  | { type: 'incrementCorrect' }
  | { type: 'incrementWrong' }
  | { type: 'reset' }

const defaultState: State = {
  correct: 0,
  wrong: 0,
  total: 0,
}

function reducer(state: State, action: Action): State {
  switch (action.type) {
    case 'incrementCorrect':
      return {
        ...state,
        correct: state.correct + 1,
        total: state.total + 1,
      }
    case 'incrementWrong':
      return {
        ...state,
        wrong: state.wrong + 1,
        total: state.total + 1,
      }
    case 'reset':
      return { ...defaultState }
    default:
      throw new Error(`Unhandled action type}`)
  }
}

const StatsStateContext = React.createContext<State | undefined>(undefined)
const StatsDispatchContext = React.createContext<Dispatch | undefined>(
  undefined
)

StatsStateContext.displayName = 'Stats'

const StatsStateConsumer = StatsStateContext.Consumer

function useStatsState(): State {
  const context = useContext(StatsStateContext)
  if (context === undefined) {
    throw new Error('useStatsState must be used within a StatsProvider')
  }
  return context
}

function useStatsDispatch(): Dispatch {
  const context = useContext(StatsDispatchContext)
  if (context === undefined) {
    throw new Error('useStatsDispatch must be used within a StatsProvider')
  }
  return context
}

function useStats(): [State, Dispatch] {
  return [useStatsState(), useStatsDispatch()]
}

const StatsProvider: React.FC<StatsProviderProps> = ({
  children,
  initialState = defaultState,
}) => {
  const [state, dispatch] = useReducer(reducer, initialState)

  return (
    <StatsStateContext.Provider value={state}>
      <StatsDispatchContext.Provider value={dispatch}>
        {children}
      </StatsDispatchContext.Provider>
    </StatsStateContext.Provider>
  )
}

export {
  reducer,
  StatsProvider,
  StatsStateConsumer,
  useStats,
  useStatsDispatch,
  useStatsState,
}
