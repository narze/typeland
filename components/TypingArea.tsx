/** @jsx jsx */
import { jsx } from '@emotion/core'
import tw from '@tailwindcssinjs/macro'
import { Word } from './Word'
import React, { useContext, useEffect } from 'react'
import { StatsContext } from '../contexts/Stats'

const s = {
  typingArea: tw`
    text-xl
    antialiased
  `,
  template: tw`
    my-1
    break-words
  `,
  user: tw`
    my-1
    break-words
  `,
}

export enum Mode {
  typealong = 'typealong',
}

export interface TypingAreaProps {
  words: Array<string>
  userWords: Array<string>
  showCaret: boolean
  mode?: Mode
  finished?: boolean
}

export const TypingArea: React.FC<TypingAreaProps> = React.memo(
  ({ words, userWords, showCaret, mode, finished }) => {
    const { stats, dispatch } = useContext(StatsContext)

    useEffect(() => {
      const offset = stats.total

      userWords.slice(offset).forEach((text, i) => {
        if (text == '') {
          return
        }

        if (text == words[offset + i]) {
          dispatch({ type: 'incrementCorrect' })
        } else {
          dispatch({ type: 'incrementWrong' })
        }
      })
    }, [userWords.length, finished])

    if (mode == 'typealong') {
      const remainingWords = words.slice(userWords.length)

      return (
        <div css={s.typingArea}>
          <span css={s.user} data-testid="user">
            {userWords.map((text, i) => {
              return (
                <React.Fragment key={i}>
                  <Word
                    template={words[i]}
                    userInput={text}
                    showCaret={showCaret && i == userWords.length - 1}
                    mode={mode}
                  />
                  <span>&nbsp;</span>
                </React.Fragment>
              )
            })}
          </span>
          <span css={s.template} data-testid="template">
            {remainingWords.join(' ')}
          </span>
        </div>
      )
    }

    return (
      <div css={s.typingArea}>
        <p css={s.template} data-testid="template">
          {words.join(' ')}
        </p>
        <p css={s.user} data-testid="user">
          {userWords.map((text, i) => {
            return (
              <React.Fragment key={i}>
                <Word
                  template={words[i]}
                  userInput={text}
                  showCaret={showCaret && i == userWords.length - 1}
                />
                <span>&nbsp;</span>
              </React.Fragment>
            )
          })}
        </p>
      </div>
    )
  }
)
