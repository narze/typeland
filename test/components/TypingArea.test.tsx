import React from 'react'
import { render } from '../testUtils'
import { TypingArea, Mode } from '@/components/TypingArea'
import { StatsContext } from '../../contexts/Stats'

const providerValue = {
  incrementCorrect: jest.fn(),
  incrementWrong: jest.fn(),
  reset: jest.fn(),
  correct: 0,
  wrong: 0,
  total: 0,
}

afterEach(() => {
  providerValue.incrementCorrect.mockClear()
  providerValue.incrementWrong.mockClear()
  providerValue.reset.mockClear()
})

const renderWithProvider = (
  ui,
  { providerProps = { value: providerValue }, ...renderOptions } = {}
) => {
  return render(
    <StatsContext.Provider {...providerProps}>{ui}</StatsContext.Provider>,
    renderOptions
  )
}

it('renders typing area', () => {
  const props = {
    words: 'the quick brown fox jumps over the lazy dog'.split(' '),
    userWords: 'the quick brown dog'.split(' '),
    showCaret: true,
  }

  const { asFragment, getByTestId } = renderWithProvider(
    <TypingArea {...props} />,
    {}
  )
  expect(getByTestId('template')).toHaveTextContent(
    'the quick brown fox jumps over the lazy dog'
  )
  expect(getByTestId('user')).toHaveTextContent('the quick brown dog')

  expect(asFragment()).toMatchSnapshot()
})

describe('typealong mode', () => {
  it('renders template & user area combined', () => {
    const props = {
      words: 'the quick brown fox jumps over the lazy dog'.split(' '),
      userWords: 'the quick brown doge'.split(' '),
      showCaret: true,
      mode: Mode.typealong,
    }

    const { asFragment, getByTestId } = renderWithProvider(
      <TypingArea {...props} />,
      {}
    )
    expect(getByTestId('user')).toHaveTextContent('the quick brown foxe')
    expect(getByTestId('template')).toHaveTextContent('jumps over the lazy dog')

    expect(asFragment()).toMatchSnapshot()
  })

  it('renders mid-word correctly', () => {
    const props = {
      words: 'the quick brown fox jumps over the lazy dog'.split(' '),
      userWords: 'the quick brown doge jum'.split(' '),
      showCaret: true,
      mode: Mode.typealong,
    }

    const { asFragment, getByTestId } = renderWithProvider(
      <TypingArea {...props} />,
      {}
    )
    expect(getByTestId('user')).toHaveTextContent('the quick brown foxe jumps')
    expect(getByTestId('template')).toHaveTextContent('over the lazy dog')

    expect(asFragment()).toMatchSnapshot()
  })

  it('exports stats via StatsContext', () => {
    const props = {
      words: 'the quick brown fox jumps over the lazy dog'.split(' '),
      userWords: 'the quick brown doge jum '.split(' '),
      showCaret: true,
      mode: Mode.typealong,
    }

    renderWithProvider(<TypingArea {...props} />)

    expect(providerValue.incrementCorrect).toHaveBeenCalledTimes(3)
    expect(providerValue.incrementWrong).toHaveBeenCalledTimes(2)

    providerValue.incrementCorrect.mockClear()
    providerValue.incrementWrong.mockClear()

    props.userWords = 'the quick brown doge jum over a lazy dog'.split(' ')

    renderWithProvider(<TypingArea {...props} />, {
      providerProps: { value: { ...providerValue, total: 5 } },
    })

    expect(providerValue.incrementCorrect).toHaveBeenCalledTimes(3)
    expect(providerValue.incrementWrong).toHaveBeenCalledTimes(1)
  })
})
