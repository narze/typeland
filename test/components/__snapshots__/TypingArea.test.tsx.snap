// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders mid-word correctly 1`] = `
<DocumentFragment>
  .emotion-33 {
  font-size: 1.25rem;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
}

.emotion-32 {
  margin-top: 0.25rem;
  margin-bottom: 0.25rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  width: 100%;
}

.emotion-3 {
  position: relative;
  margin-right: 0.25rem;
}

.emotion-0 {
  --text-opacity: 1;
  color: #057a55;
  color: rgba(5,122,85,var(--text-opacity));
}

.emotion-16 {
  --text-opacity: 1;
  color: #f98080;
  color: rgba(249,128,128,var(--text-opacity));
}

.emotion-28 {
  margin-right: 0.25rem;
}

.emotion-26 {
  --bg-opacity: 1;
  position: absolute;
  height: 1.5rem;
  background-color: #3f83f8;
  background-color: rgba(63,131,248,var(--bg-opacity));
  -webkit-transition: all 0.1s ease-out;
  transition: all 0.1s ease-out;
  width: 2px;
  left: 1px;
  top: 4px;
}

.emotion-24 {
  --text-opacity: 1;
  color: #d2d6dc;
  color: rgba(210,214,220,var(--text-opacity));
}

<div
    class="emotion-33"
    data-testid="text"
  >
    <div
      class="emotion-32"
    >
      <span
        class="emotion-3"
        data-testid="word"
      >
        <span
          class="correct emotion-0"
          data-testid="correct"
        >
          t
        </span>
        <span
          class="correct emotion-0"
          data-testid="correct"
        >
          h
        </span>
        <span
          class="correct emotion-0"
          data-testid="correct"
        >
          e
        </span>
      </span>
      <span
        class="emotion-3"
        data-testid="word"
      >
        <span
          class="correct emotion-0"
          data-testid="correct"
        >
          q
        </span>
        <span
          class="correct emotion-0"
          data-testid="correct"
        >
          u
        </span>
        <span
          class="correct emotion-0"
          data-testid="correct"
        >
          i
        </span>
        <span
          class="correct emotion-0"
          data-testid="correct"
        >
          c
        </span>
        <span
          class="correct emotion-0"
          data-testid="correct"
        >
          k
        </span>
      </span>
      <span
        class="emotion-3"
        data-testid="word"
      >
        <span
          class="correct emotion-0"
          data-testid="correct"
        >
          b
        </span>
        <span
          class="correct emotion-0"
          data-testid="correct"
        >
          r
        </span>
        <span
          class="correct emotion-0"
          data-testid="correct"
        >
          o
        </span>
        <span
          class="correct emotion-0"
          data-testid="correct"
        >
          w
        </span>
        <span
          class="correct emotion-0"
          data-testid="correct"
        >
          n
        </span>
      </span>
      <span
        class="emotion-3"
        data-testid="word"
      >
        <span
          class="wrong emotion-16"
          data-testid="wrong"
        >
          f
        </span>
        <span
          class="correct emotion-0"
          data-testid="correct"
        >
          o
        </span>
        <span
          class="wrong emotion-16"
          data-testid="wrong"
        >
          x
        </span>
        <span
          class="wrong emotion-16"
          data-testid="wrong"
        >
          e
        </span>
      </span>
      <span
        class="emotion-3"
        data-testid="word"
      >
        <span
          class="correct emotion-0"
          data-testid="correct"
        >
          j
        </span>
        <span
          class="correct emotion-0"
          data-testid="correct"
        >
          u
        </span>
        <span
          class="correct emotion-0"
          data-testid="correct"
        >
          m
        </span>
        <span
          class="pending emotion-24"
          data-testid="pending"
        >
          p
        </span>
        <span
          class="pending emotion-24"
          data-testid="pending"
        >
          s
        </span>
        <span
          class="emotion-26"
          data-testid="caret"
        />
      </span>
      <span
        class="emotion-28"
        data-testid="remainingWord"
      >
        over
      </span>
      <span
        class="emotion-28"
        data-testid="remainingWord"
      >
        the
      </span>
      <span
        class="emotion-28"
        data-testid="remainingWord"
      >
        lazy
      </span>
      <span
        class="emotion-28"
        data-testid="remainingWord"
      >
        dog
      </span>
    </div>
  </div>
</DocumentFragment>
`;

exports[`renders typing area 1`] = `
<DocumentFragment>
  .emotion-27 {
  font-size: 1.25rem;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
}

.emotion-26 {
  margin-top: 0.25rem;
  margin-bottom: 0.25rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  width: 100%;
}

.emotion-3 {
  position: relative;
  margin-right: 0.25rem;
}

.emotion-0 {
  --text-opacity: 1;
  color: #057a55;
  color: rgba(5,122,85,var(--text-opacity));
}

.emotion-16 {
  --text-opacity: 1;
  color: #f98080;
  color: rgba(249,128,128,var(--text-opacity));
}

.emotion-21 {
  margin-right: 0.25rem;
}

.emotion-19 {
  --bg-opacity: 1;
  position: absolute;
  height: 1.5rem;
  background-color: #3f83f8;
  background-color: rgba(63,131,248,var(--bg-opacity));
  -webkit-transition: all 0.1s ease-out;
  transition: all 0.1s ease-out;
  width: 2px;
  left: 1px;
  top: 4px;
}

<div
    class="emotion-27"
    data-testid="text"
  >
    <div
      class="emotion-26"
    >
      <span
        class="emotion-3"
        data-testid="word"
      >
        <span
          class="correct emotion-0"
          data-testid="correct"
        >
          t
        </span>
        <span
          class="correct emotion-0"
          data-testid="correct"
        >
          h
        </span>
        <span
          class="correct emotion-0"
          data-testid="correct"
        >
          e
        </span>
      </span>
      <span
        class="emotion-3"
        data-testid="word"
      >
        <span
          class="correct emotion-0"
          data-testid="correct"
        >
          q
        </span>
        <span
          class="correct emotion-0"
          data-testid="correct"
        >
          u
        </span>
        <span
          class="correct emotion-0"
          data-testid="correct"
        >
          i
        </span>
        <span
          class="correct emotion-0"
          data-testid="correct"
        >
          c
        </span>
        <span
          class="correct emotion-0"
          data-testid="correct"
        >
          k
        </span>
      </span>
      <span
        class="emotion-3"
        data-testid="word"
      >
        <span
          class="correct emotion-0"
          data-testid="correct"
        >
          b
        </span>
        <span
          class="correct emotion-0"
          data-testid="correct"
        >
          r
        </span>
        <span
          class="correct emotion-0"
          data-testid="correct"
        >
          o
        </span>
        <span
          class="correct emotion-0"
          data-testid="correct"
        >
          w
        </span>
        <span
          class="correct emotion-0"
          data-testid="correct"
        >
          n
        </span>
      </span>
      <span
        class="emotion-3"
        data-testid="word"
      >
        <span
          class="wrong emotion-16"
          data-testid="wrong"
        >
          f
        </span>
        <span
          class="correct emotion-0"
          data-testid="correct"
        >
          o
        </span>
        <span
          class="wrong emotion-16"
          data-testid="wrong"
        >
          x
        </span>
        <span
          class="emotion-19"
          data-testid="caret"
        />
      </span>
      <span
        class="emotion-21"
        data-testid="remainingWord"
      >
        jumps
      </span>
      <span
        class="emotion-21"
        data-testid="remainingWord"
      >
        over
      </span>
      <span
        class="emotion-21"
        data-testid="remainingWord"
      >
        the
      </span>
      <span
        class="emotion-21"
        data-testid="remainingWord"
      >
        lazy
      </span>
      <span
        class="emotion-21"
        data-testid="remainingWord"
      >
        dog
      </span>
    </div>
  </div>
</DocumentFragment>
`;
